### ISSUE TITLE
change the default prior to bounded prior

### ISSUE THREAD

**Issue by tliu1023** [AdamCobb/hamiltorch#14]
Hi Adam,

Thanks for this great repo! I'm new to bayesian and the repo helps me a lot. Now I'm trying to use the hamiltorch NUTS in my project. But I'm trying to use a different prior distribution (ex. uniform distribution, gaussian without zero mean). But the issue is that when I use a uniform distribution, even though I have a very loose bound (lb=-100, ub=100) and very small step size(1e-3), but in leapfrog, the parameter still walks beyond the bound and then cause an error when calculating log_prior. I'm wondering is there a good way to use a bounded prior? Thanks for your time and help!

Best,
Tong

**Comment by AdamCobb** [comment]
Hi Tong,
 
Thanks for reaching out! I think it is cool that you are looking into adding different priors – definitely a worthy topic for exploration.
 
I think that using bounded priors does risk leaving the support of the prior, as HMC naively does not know where the boundary lies. Also a uniform distribution will not “discourage” your sample from leaving the support of until it is too late. What I mean by that is that at least with a Gaussian prior, we are pushing the sampler to be around zero and that going to say a value of -10 for a N(0,1) would be extremely unlikely. A uniform prior of U(-100,100) would mean being in the location of 0 is just as likely as 100. This will cause an issue if your sampler gets close to the boundary.
 
If you use simple HMC with no adaptive step-size, then you may be lucky and stay around 0 or within the support. If you adapt the step-size using NUTS (i.e. the implementation in hamiltorch), then you may jump around a lot at the beginning and end up near the boundary of the uniform prior. This possibly leads to your issues.
 
If you really want to use a bounded prior, you will have to catch errors with a “try and except” and just reject any of those proposals as they have zero probability.
 
All the best,
 
Adam

**Comment by tliu1023** [comment]
Hi Adam,

Thanks for replying! And thanks for explaining what is going on under the hood of HMC and NUTS. Your answer makes more sense to me! That might be the reason why I got this error.

'try and expect' syntax is exactly what I did right now. And also thanks for point out the fact that it might be going too close to the boundary until it turns back. I didn't think about it before. And you are right might be uniform pdf is too distributed than gaussian pdf. It might be a better idea if use truncated normal in bounded case. Thank you!

Best,
Tong
